source/common/http/http1/codec_impl.cc

```diff
@@ -411,8 +411,9 @@ int ServerConnectionImpl::onHeadersComplete(HeaderMapImplPtr&& headers) {
         http_parser_pause(&parser_, 1);
       }
 
-    } else if (parser_.method == http_method::HTTP_POST ||
-               parser_.method == http_method::HTTP_PUT) {
+    } else if ((parser_.method == http_method::HTTP_POST ||
+                parser_.method == http_method::HTTP_PUT) &&
+               headers->ContentLength() == nullptr) {
```

Comment:
I guess this comes back to my question before on what http-parser does if content-length is omitted and there is no chunk encoding. IIRC it infers content length is 0 during parsing.